Add support for mdi images

guillempages opened the pull request:
# What does this implement/fix?

This PR add support for specifying Material Design Icons directly in the YAML configuration, without the need of manually downloading them and converting them to PNG.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** N/A

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#2837

## Test Environment

- [x] ESP32
- [ ] ESP32 IDF
- [ ] ESP8266
- [ ] RP2040

## Example entry for `config.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

```yaml
# Example config.yaml
image:
  - file: mdi:alert-outline
    resize: 80x80
    id: alert
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

guillempages committed:
Add support for direct MDI icons usage

guillempages commented on esphome/components/image/__init__.py:
```diff
@@ -50,86 +59,180 @@
+                config[CONF_ID],
+                config[CONF_TYPE],
+            )
+        from pathlib import Path
+        import requests
+        from svglib.svglib import svg2rlg
+        from reportlab.graphics import renderPM
```
Is this the right place to put the imports (like the PIL import in line 60)?
Or would it be better to put them at the top of the file?

guillempages replied:
Moved them to the top of the file

guillempages commented on esphome/components/image/__init__.py:
```diff
@@ -22,21 +22,30 @@
+        cv.Exclusive(CONF_FILE, "input"): cv.file_,
+        cv.Exclusive(CONF_MDI, "input"): cv.string,
         cv.Optional(CONF_RESIZE): cv.dimensions,
         cv.Optional(CONF_TYPE, default="BINARY"): cv.enum(IMAGE_TYPE, upper=True),
```
For MDI images, "TRANSPARENT_BINARY" would be a better default; need to find a way of having sensible defaults depending on the rest of the parameters.

guillempages replied:
Decided to force the type to be TRANSPARENT_BINARY for MDI images; everything else does not make any sense IMHO

guillempages commented on esphome/components/image/__init__.py:
```diff
@@ -50,86 +59,180 @@
+
+        mdi = config[CONF_MDI]
+        url = f"https://raw.githubusercontent.com/Templarian/MaterialDesign/master/svg/{mdi}.svg"
+        _LOGGER.debug("Downloading %s MDI image from %s", mdi, url)
```
Should this be "_LOGGER.info" instead of debug? Probably not, since in case of error the URL is printed anyway.

guillempages replied:
Changed to info, since the image is now only downloaded once every 24 hours.

guillempages commented:
The PR is made on top of #4600; once that one is merged I can rebase to make the modification smaller / the review easier

guillempages committed:
Add tests for mdi images

guillempages committed:
Allow using intellisense code completion for MDI images

Unfocused commented on requirements_optional.txt:
```diff
@@ -1,2 +1,4 @@
 pillow>4.0.0
+svglib>=1.5.0
+reportlab>=3.5.0
```
Will need to not just document this, but also add a runtime check & warning if they're missing (instead of unconditionally importing them).

At the moment, if these aren't installed, the `image`, `animation`, and `online_image` components effectively disappear - ESPHome exits with an error that those components can't be found if a device config tries to use them (not the most helpful message, admittedly). `animation` and `online_image` are affected despite not using these new libraries because they both import from the `image` component, which of course breaks them too.

See here for how the `font` component handles needing Pillow:
https://github.com/esphome/esphome/blob/4b664b6f09a6eef16e380a953973eeda59ea2f19/esphome/components/font/__init__.py#L62-L77

Ideally they'd be checked & imported only if they're actually going to be used - if someone never uses MDI, this shouldn't get in their way.

guillempages replied:
I actually asked @jesserockz what's the official policy with imports, and the answer was "imports at the beginning unless there is a reason not to", so that's why I put that at the beginning.

That said, I agree that it's not nice to crash silently if those are not installed; especially when not used. So I'll revisit that.

guillempages committed:
Conditionally import svglib and dependencies

If mdi images are not used, there is no need to use the additional
new dependencies (svglib and reportlab), so they should also not
be automatically included in those cases.

guillempages committed:
Support for "file: mdi:xyz" format

Instead of having separate "CONF_FILE" and "CONF_MDI", just use "CONF_FILE" with an "mdi:" prefix to detect mdi images.

guillempages commented on esphome/components/image/__init__.py:
```diff
@@ -78,8 +85,10 @@ def validate_cross_dependencies(config):
     have "use_transparency" set to True.
     Also set the default value for those kind of dependent fields.
     """
+    is_mdi = CONF_FILE in config and config[CONF_FILE][CONF_SOURCE] == SOURCE_MDI
```
FYI, the online runtime images commit I am working on, also uses this method, but does not have `CONF_FILE`, it has `CONF_URL` instead. That's why I'm explicitly checking for `CONF_FILE in config` even though `CONF_FILE` is Required.

Unfocused replied:
Really like this change to use `file: "mdi:whatever"` instead. 

I was debating whether to suggest something like that, since using `mdi` as the key name meant tightly coupling icon support to explicitly only MDI, which didn't feel future-proof.

guillempages committed:
Use cairosvg instead of reportlab
